control/server.py

```diff
@@ -0,0 +1,279 @@
+            server_key = self.settings.get("mtls", "server_key")
+            server_cert = self.settings.get("mtls", "server_cert")
+            client_cert = self.settings.get("mtls", "client_cert")
+            with open(server_key, "rb") as f:
+                private_key = f.read()
+            with open(server_cert, "rb") as f:
+                server_crt = f.read()
+            with open(client_cert, "rb") as f:
+                client_crt = f.read()
+
+            # Create appropriate server credentials
+            server_credentials = grpc.ssl_server_credentials(
+                private_key_certificate_chain_pairs=[(private_key, server_crt)],
+                root_certificates=client_crt,
+                require_client_auth=True,
+            )
+
+            # Add secure port using crendentials
+            self.server.add_secure_port(
+                "{}:{}".format(gateway_addr, gateway_port), server_credentials)
+        else:
+            # Authentication is not enabled
+            self.server.add_insecure_port("{}:{}".format(
+                gateway_addr, gateway_port))
+
+        # Start server
+        self.logger.info("Starting server")
+        self.server.start()
+        # Check for existing NVMeoF target configuration
+        self._restore_config()
+        # Start checking for persistent configuration updates
+        self.update_timer = threading.Thread(target=self._update_config)
+        self.update_timer.daemon = True
+        self.update_timer.start()
+
+        while True:
+            timedout = self.server.wait_for_termination(timeout=1)
+            if not timedout:
+                break
+            alive = self._ping()
+            if not alive:
+                break
+
+    def _restore_config(self):
+        """Restores NVMeoF target configuration."""
+
+        callbacks = {
+            self.persistent_config.BDEV_PREFIX: {
+                True: self.gateway_rpc.create_bdev,
+                False: self.gateway_rpc.delete_bdev
+            },
+            self.persistent_config.SUBSYSTEM_PREFIX: {
+                True: self.gateway_rpc.create_subsystem,
+                False: self.gateway_rpc.delete_subsystem
+            },
+            self.persistent_config.NAMESPACE_PREFIX: {
+                True: self.gateway_rpc.add_namespace,
+                False: self.gateway_rpc.remove_namespace
+            },
+            self.persistent_config.HOST_PREFIX: {
+                True: self.gateway_rpc.add_host,
+                False: self.gateway_rpc.remove_host
+            },
+            self.persistent_config.LISTENER_PREFIX: {
+                True: self.gateway_rpc.add_listener,
+                False: self.gateway_rpc.remove_listener
+            }
+        }
+        self.persistent_config.restore(callbacks)
+
+    def _update_config(self):
+        """Periodically check for persistent config updates."""
+        update_interval = self.settings.getfloat("gateway", "refresh")
```

Comment:
Let's not use a float value in the user provided config to specify an interval. We need a defined time unit and rename the config parameter accordingly, e.g. `config_update_interval_sec` and have a way to turn it of e.g. by setting it to 0 or -1.